manual/src/tutorials/memorymodel.etex

```diff
@@ -0,0 +1,790 @@
+consistent behaviour.
+
+Since "s" and "o" are newly allocated in "d1", the only schedule where "d2"
+will read those objects is the one in which "d1" executes before "d2". Thanks
+to LDRF-SC, we can use sequential reasoning, which guarantees that "d2" will
+always see the initialising writes from "d1".
+
+\section{s:mm_semantics}{An operational view of the memory model}
+
+In this section, we describe the semantics of the OCaml memory model. Given an
+OCaml program, which may possibly contain data races, the semantics tells you
+the values that may be observed by the read of a memory location. For
+simplicity, we restrict the intra-thread actions to just the accesses to atomic
+and non-atomic locations, ignoring domain spawn and join operations, and the
+operations on mutexes.
+
+We describe the semantics of the OCaml memory model in a straightforward
+small-step operational manner. That is, the semantics is described by an
+abstract machine that executes one action at a time, arbitrarily picking one of
+the available domains at each step. This is similar to the abstract machine
+that we had used to describe the happens-before relationship in
+section~\ref{s:happens_before}.
+
+\subsection{ss:mm_non_atomic}{Non-atomic locations}
+
+In the semantics, we model non-atomic locations as a finite map from timestamps
+"t" to values "v". We take timestamps to be rational numbers. The timestamps
+are totally ordered but dense; there is a timestamp between any two others.
+
+For example,
+
+\begin{verbatim}
+a: [t1 -> 1; t2 -> 2]
+b: [t3 -> 3; t4 -> 4; t5 -> 5]
+c: [t6 -> 5; t7 -> 6; t8 -> 7]
+\end{verbatim}
+
+\noindent represents three non-atomic locations "a", "b" and "c" and their
+histories. The location "a" has two writes at timestamps "t1" and "t2" with
+values "1" and "2", respectively. When we write "a: [t1 -> 1; t2 -> 2]", we
+assume that "t1 < t2". We assume that the locations are initialised with a
+history that has a single entry at timestamp 0 that maps to the initial value.
+
+\subsection{ss:mm_domains}{Domains}
+
+Each domain is equipped with a \emph{frontier}, which is a map from non-atomic
+locations to timestamps. Intuitively, each domain's frontier records, for each
+non-atomic location, the latest write known to the thread. More recent writes
+may have occurred, but are not guaranteed to be visible.
+
+For example,
+
+\begin{verbatim}
+d1: [a -> t1; b -> t3; c -> t6]
+d2: [a -> t1; b -> t4; c -> t7]
+\end{verbatim}
+
+\noindent represents two domains "d1" and "d2" and their frontiers.
+
+\subsection{ss:mm_na_access}{Non-atomic accesses}
+
+Let us now define the semantics of non-atomic reads and writes. Suppose domain
+"d1" performs the read of "b". For non-atomic reads, the domains may read an
+arbitrary element of the history for that location, as long as it is not older
+than the timestamp in the domains's frontier. In this case, since "d1" frontier
+at "b" is at "t3", the read may return the value "3", "4" or "5".
```

Comment:
Add a comment saying that a non-atomic read does not change the frontier of the current domain. This is far from obvious.

Reply:
Done.